src/main/java/com/mongodb/spark/sql/connector/schema/RowToBsonDocumentConverter.java

```diff
@@ -105,54 +113,86 @@ public BsonDocument fromRow(final InternalRow row) {
-        return BsonNull.VALUE;
-      } else if (dataType instanceof DecimalType) {
+      }
+      try {
+        return cachedObjectToBsonValue.apply(data);
+      } catch (Exception e) {
+        throw new DataException(format(
+            "Cannot cast %s into a BsonValue. %s has no matching BsonValue. Error: %s",
+            data, dataType, e.getMessage()));
+      }
+    };
+  }
+
+  private static ObjectToBsonElement dataTypeToBsonElement(
+      final StructField field,
+      final WriteConfig.ConvertJson convertJson,
+      final boolean ignoreNulls) {
+    ObjectToBsonValue cachedObjectToBsonValue =
+        createObjectToBsonValue(field.dataType(), convertJson, ignoreNulls);
+    return (data) -> new BsonElement(field.name(), cachedObjectToBsonValue.apply(data));
+  }
+
+  @SuppressWarnings("unchecked")
+  private static ObjectToBsonValue objectToBsonValue(
+      final DataType dataType,
+      final WriteConfig.ConvertJson convertJson,
+      final boolean ignoreNulls) {
+    if (DataTypes.BinaryType.acceptsType(dataType)) {
+      return (data) -> new BsonBinary((byte[]) data);
+    } else if (DataTypes.BooleanType.acceptsType(dataType)) {
+      return (data) -> new BsonBoolean((Boolean) data);
+    } else if (DataTypes.DoubleType.acceptsType(dataType)) {
+      return (data) -> new BsonDouble(((Number) data).doubleValue());
+    } else if (DataTypes.FloatType.acceptsType(dataType)) {
+      return (data) -> new BsonDouble(((Number) data).floatValue());
+    } else if (DataTypes.IntegerType.acceptsType(dataType)) {
+      return (data) -> new BsonInt32(((Number) data).intValue());
+    } else if (DataTypes.ShortType.acceptsType(dataType)) {
+      return (data) -> new BsonInt32(((Number) data).intValue());
+    } else if (DataTypes.ByteType.acceptsType(dataType)) {
+      return (data) -> new BsonInt32(((Number) data).intValue());
+    } else if (DataTypes.LongType.acceptsType(dataType)) {
+      return (data) -> new BsonInt64(((Number) data).longValue());
+    } else if (DataTypes.StringType.acceptsType(dataType)) {
+      return (data) -> processString((String) data, convertJson);
+    } else if (DataTypes.DateType.acceptsType(dataType)
+        || DataTypes.TimestampType.acceptsType(dataType)) {
+      return (data) -> new BsonDateTime(((Date) data).getTime());
+    } else if (DataTypes.NullType.acceptsType(dataType)) {
+      return (data) -> BsonNull.VALUE;
+    } else if (dataType instanceof DecimalType) {
+      return (data) -> {
         BigDecimal bigDecimal = data instanceof BigDecimal
             ? (BigDecimal) data
             : ((Decimal) data).toBigDecimal().bigDecimal();
         return new BsonDecimal128(new Decimal128(bigDecimal));
-      } else if (dataType instanceof ArrayType) {
+      };
+    } else if (dataType instanceof ArrayType) {
+      return (data) -> {
         DataType elementType = ((ArrayType) dataType).elementType();
+        Function<Object, BsonValue> element =
```

Comment:
To avoid a performance penalty here, should this be lifted out of the lambda?

Reply:
Done